// Copyright 2025 The Ray Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

import "src/ray/protobuf/runtime_env_common.proto";
import "src/ray/protobuf/common.proto";

package ray.rpc.events;

// Message containing the definition information of a task.
// The message is expected to be emitted once per task attempt.
message TaskDefinitionEvent {
  // task_id and task_attempt form the unique identifier of a task.
  bytes task_id = 1;
  int32 task_attempt = 2;

  // The task definition information.
  // Valid values are NORMAL_TASK, ACTOR_CREATION_TASK, DRIVER_TASK
  TaskType task_type = 3;
  Language language = 4;
  FunctionDescriptor task_func = 5;
  string task_name = 6;
  map<string, double> required_resources = 7;
  RuntimeEnvInfo runtime_env_info = 8;

  // The correlation ids of the task that can be used to correlate the task with
  // other events.
  bytes job_id = 9;
  bytes parent_task_id = 10;
  bytes placement_group_id = 11;
  map<string, bytes> ref_ids = 12;
}
